walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,181 @@
+    if (attendeesList.size() > 0 && availabilityMap.size() == 0) {
+      return null;
+    }
+
+    ArrayList<ArrayList<TimeRange>> freeTimeList =
+        new ArrayList<ArrayList<TimeRange>>(availabilityMap.values());
+    ArrayList<TimeRange> finalList = getAvailable(freeTimeList, request.getDuration());
+    System.out.println(finalList);
+
+    return finalList;
+  }
+
+  /**
+   * Compares time ranges and returns a time range of the overlap. Boolean nextPerson checks whether
+   * there is another person to compare time ranges with.
+   */
+  private ArrayList<TimeRange> getAvailable(
+      ArrayList<ArrayList<TimeRange>> allFreeTimeList, long requestDuration) {
+    ArrayList<TimeRange> availableOverlapList = new ArrayList<TimeRange>();
+    Integer allFreeTimeListSize = new Integer(allFreeTimeList.size());
+
+    // Validates the list is not empty.
+    if (allFreeTimeListSize.equals(0)) {
+      return null;
+    }
+
+    // If only one person is requestd, return their free time.
+    if (allFreeTimeListSize.equals(1)) {
+      return allFreeTimeList.get(0);
+    }
+
+    // Compares the first 2 TimeRange in freeTimeList and returns the overlap if only 2
+    // elements in freeTimeList or calls getAvailable recursively by replacing the first 2
+    // TimeRanges with their overlapped TimeRange
+    ArrayList<TimeRange> freeTimeList1 = allFreeTimeList.get(0);
+    ArrayList<TimeRange> freeTimeList2 = allFreeTimeList.get(1);
+
+    // Compares each time range for the second person to the time ranges of the first person.
+    for (TimeRange range1 : freeTimeList1) {
+      for (TimeRange range2 : freeTimeList2) {
+
+        // Checks if there is an overlap between the 2 ranges.
+        if (range1.overlaps(range2)) {
+          long overlapStart = Math.max(range1.start(), range2.start());
+          long overlapEnd = Math.min(range1.end(), range2.end());
+          long overlapDuration = overlapEnd - overlapStart;
+
+          // Verifies that the overlap duration is greater than or equal to request duration.
+          if (overlapDuration >= requestDuration) {
+            TimeRange overlapTimeRange =
+                TimeRange.fromStartDuration((int) overlapStart, (int) overlapDuration);
+                
+            // If there are more requested attendees in the allFreeTimeList, must compare their
+            // free time with the overlapped free time of previous requested attendees.
+            if (allFreeTimeListSize > 2) {
+              ArrayList<TimeRange> overlapArray = new ArrayList<TimeRange>();
+              overlapArray.add(overlapTimeRange);
+              ArrayList<ArrayList<TimeRange>> newAllFreeTimeList =
+                  new ArrayList<ArrayList<TimeRange>>(
+                      allFreeTimeList.subList(2, allFreeTimeListSize));
+              newAllFreeTimeList.add(0, overlapArray);
+              availableOverlapList.addAll(getAvailable(newAllFreeTimeList, requestDuration));
```

Comment:
Overall, the algo worked and the idea was interesting. However, the recursion here is the bottle neck of this algorithm (and its time complexity is super high which is `O(N^N)` and `N` is the number of time ranges).

I would recommend the following readings (https://www.geeksforgeeks.org/merging-intervals/) and here are my thoughts:
1. Since we have all the unavailable time ranges of these attendees, we should look for a way to merge all of them. 
2. we can sort the unavailable timerangs first by each timerange's start time. (which will take `O(NlgN)` time, and here `N` is the num of time ranges).
3. then we could use a stack (or other data structure) to merge the sorted timeranges in linear time. (`O(N)` time)
4. after getting the merged unavailable time ranges, we can then loop through them and figure out available time ranges. (`O(N)` here N s the num of unavailable time ranges).

The time complexity will be improved by this sorting-merging-looping solution. We can definitely talk more about it tomorrow during our 1:1.